### PR TITLE
Optimize multiple filters call

### PR DESCRIPTION
**Pull request by ganmacs**
**Please merge this PR after #1140 is merged, Becasue this PR depends on https://github.com/fluent/fluentd/pull/1140.**

If your code can be optimized, the speed of `filter_stream` will be about 1.2 times faster.
This result is caused by avoiding to call a lot of `Fluent::MultiEventStream#add`.

the requirement of running optimization is that filter plugins that you are using don't implement `filter_stream`.

the executed command is here.
```
bundle exec fluentd -c example/multi_filters.conf
```

The measured code, as shown below. (https://github.com/fluent/fluentd/pull/1145/files#diff-c18beef6caee95367d6268a4c181f913R166)

```rb
def filter_stream(tag, es)
  require 'ruby-prof'
  RubyProf.start

  if optimizable?
    processed = optimized_filter_stream(tag, es)
  else
    processed = @filters.reduce(es) { |acc, filter| filter.filter_stream(tag, acc) }
  end

  result = RubyProf.stop
  printer = RubyProf::FlatPrinter.new(result)
  printer.print(STDOUT)
  processed
end
```

The result o calling `optimized_filter_stream`

```
* indicates recursively called methods
Measure Mode: wall_time
Thread ID: 70252402759920
Fiber ID: 70252404764700
Total: 0.051646
Sort by: self_time

 %self      total      self      wait     child     calls  name
 19.05      0.024     0.010     0.000     0.014    20000   Hash#each
 15.70      0.041     0.008     0.000     0.033    10000   Fluent::Plugin::GrepFilter#filter
 11.29      0.006     0.006     0.000     0.000    10000   Regexp#match
 10.95      0.011     0.006     0.000     0.006    10000   <Module::Fluent::Compat::StringUtil>#match_regexp
  4.66      0.002     0.002     0.000     0.000    10000   NilClass#to_s
  1.78      0.051     0.001     0.000     0.050    11000  *Kernel#catch
  1.29      0.052     0.001     0.000     0.051     1001  *Array#each
  1.15      0.001     0.001     0.000     0.000     1000   Fluent::MultiEventStream#add
  0.02      0.052     0.000     0.000     0.052        1   Fluent::EventRouter::Pipeline::FilterOptimizer#optimized_filter_stream
  0.02      0.052     0.000     0.000     0.052        1   Fluent::EventRouter::Pipeline::FilterOptimizer#filter_stream
  0.01      0.000     0.000     0.000     0.000        1   Fluent::MultiEventStream#initialize
  0.01      0.000     0.000     0.000     0.000        1   Class#new
  0.00      0.052     0.000     0.000     0.052        1   Fluent::ArrayEventStream#each
```

This result of calling `@filters.reduce(es) { |acc, filter| filter.filter_stream(tag, acc) }`

```
* indicates recursively called methods
Measure Mode: wall_time
Thread ID: 70231922124020
Fiber ID: 70231922694520
Total: 0.063186
Sort by: self_time

 %self      total      self      wait     child     calls  name
 17.57      0.026     0.011     0.000     0.015    20000   Hash#each
 17.14      0.037     0.011     0.000     0.026    10000   Kernel#catch
 16.16      0.057     0.010     0.000     0.047        9   Range#each
 14.94      0.046     0.009     0.000     0.037    10000   Fluent::Plugin::GrepFilter#filter
 10.08      0.012     0.006     0.000     0.006    10000   <Module::Fluent::Compat::StringUtil>#match_regexp
  9.56      0.006     0.006     0.000     0.000    10000   Regexp#match
  8.69      0.005     0.005     0.000     0.000    10000   Fluent::MultiEventStream#add
  4.23      0.003     0.003     0.000     0.000    10000   NilClass#to_s
  0.04      0.063     0.000     0.000     0.063       10   Fluent::Plugin::Filter#filter_stream
  0.03      0.000     0.000     0.000     0.000       10   Class#new
  0.03      0.057     0.000     0.000     0.057        9   Fluent::MultiEventStream#each
  0.02      0.063     0.000     0.000     0.063        2  *Array#each
  0.02      0.000     0.000     0.000     0.000       10   Fluent::MultiEventStream#initialize
  0.02      0.063     0.000     0.000     0.063        1   Fluent::EventRouter::Pipeline::FilterOptimizer#filter_stream
  0.01      0.006     0.000     0.000     0.006        1   Fluent::ArrayEventStream#each
  0.01      0.063     0.000     0.000     0.063        1   Enumerable#reduce

```

### Environment

* Tool : https://github.com/ruby-prof/ruby-prof
* OS: 
```
ProductName:    Mac OS X
ProductVersion: 10.11.6
BuildVersion:   15G31
```
* PROCESSOR: `2.7 GHz Intel Core i5`
* MEMORY: `8 GB 1867 MHz DDR3`

### CODE REVIEW

**Comment by repeatedly** [comment]
You should also measure one filter case.

File: lib/fluent/event_router.rb [hunk 1]
```diff
@@ -136,23 +136,85 @@ class Pipeline
...
+        def filters=(filters)
+          @filters = filters
+          reset_optimization
+        end
+
+        def filter_stream(tag, es)
+          if optimizable?
+            optimized_filter_stream(tag, es)
+          else
+            $log.info "Can't optimized filters, because some plugins implement `filter_stream`"
```
Comment (tagomoris) [inline lib/fluent/event_router.rb]: Log messages should be for users, not only for plugin developers.
So it's better to show "filtering works with worse performance, 'xxx_filter' users #filter_stream method" or so.
Users will get known about performance, and can complaint on that plugin about implementation.

**Comment by ganmacs** [comment]
@repeatedly 
The result of one filter case is below.

The result of calling `optimized_filter_stream`
```
* indicates recursively called methods
Measure Mode: wall_time
Thread ID: 70314214842120
Fiber ID: 70314197783380
Total: 0.007549
Sort by: self_time

 %self      total      self      wait     child     calls  name
 19.04      0.001     0.001     0.000     0.000     1000   Regexp#match
 16.03      0.005     0.001     0.000     0.004     1000   Kernel#catch
 15.85      0.004     0.001     0.000     0.002     2000   Hash#each
 13.13      0.006     0.001     0.000     0.005     1000   Fluent::Plugin::GrepFilter#filter
 10.82      0.001     0.001     0.000     0.000     1000   Fluent::MultiEventStream#add
  8.70      0.002     0.001     0.000     0.001     1000   <Module::Fluent::Compat::StringUtil>#match_regexp
  3.63      0.000     0.000     0.000     0.000     1000   NilClass#to_s
  0.16      0.008     0.000     0.000     0.008        1   Fluent::EventRouter::Pipeline::FilterOptimizer#filter_stream
  0.09      0.008     0.000     0.000     0.008        1   Fluent::Plugin::Filter#filter_stream
  0.07      0.008     0.000     0.000     0.008        2  *Array#each
  0.07      0.008     0.000     0.000     0.008        1   Fluent::ArrayEventStream#each
  0.05      0.000     0.000     0.000     0.000        1   Fluent::MultiEventStream#initialize
  0.04      0.008     0.000     0.000     0.008        1   Enumerable#reduce
  0.04      0.000     0.000     0.000     0.000        1   Class#new
```


The result of calling @filters.reduce(es) { |acc, filter| filter.filter_stream(tag, acc) }
```
* indicates recursively called methods
Measure Mode: wall_time
Thread ID: 70182152853480
Fiber ID: 70182152439780
Total: 0.008376
Sort by: self_time

 %self      total      self      wait     child     calls  name
 14.62      0.003     0.001     0.000     0.002     2000   Hash#each
 11.96      0.008     0.001     0.000     0.007     2000  *Kernel#catch
 10.94      0.005     0.001     0.000     0.004     1000   Fluent::Plugin::GrepFilter#filter
  8.15      0.008     0.001     0.000     0.008     1001  *Array#each
  8.12      0.001     0.001     0.000     0.001     1000   <Module::Fluent::Compat::StringUtil>#match_regexp
  7.74      0.001     0.001     0.000     0.000     1000   Regexp#match
  6.93      0.001     0.001     0.000     0.000     1000   Fluent::MultiEventStream#add
  4.40      0.000     0.000     0.000     0.000     1000   NilClass#to_s
  0.12      0.008     0.000     0.000     0.008        1   Fluent::EventRouter::Pipeline::FilterOptimizer#filter_stream
  0.07      0.008     0.000     0.000     0.008        1   Fluent::EventRouter::Pipeline::FilterOptimizer#optimized_filter_stream
  0.05      0.000     0.000     0.000     0.000        1   Class#new
  0.04      0.008     0.000     0.000     0.008        1   Fluent::ArrayEventStream#each
  0.04      0.000     0.000     0.000     0.000        1   Fluent::MultiEventStream#initialize
```


this is configure file.
```
<source>
  @type dummy
  tag   test
  size 1000
</source>

<filter test>
  @type grep
  exclude1 hello .
</filter>

<match test>
  @type buffered_null
</match>
```

**Comment by tagomoris** [comment]
#1140 was merged into master.
@ganmacs Could you rebase for detailed code review?

File: lib/fluent/event_router.rb [hunk 2]
```diff
@@ -136,23 +136,94 @@ class Pipeline
...
+          new_es = MultiEventStream.new
+          es.each do |time, record|
+            filtered_record = record
+            filtered_time = time
+
+            catch :break_loop do
+              @filters.each do |filter|
+                if filter.has_filter_with_time
+                  begin
+                    filtered_time, filtered_record = *filter.filter_with_time(tag, filtered_time, filtered_record)
```
Comment (tagomoris) [inline lib/fluent/event_router.rb]: There are no need to use splat (*) to return 2 values from method

File: test/test_event_router.rb [hunk 3]
```diff
@@ -184,9 +184,25 @@ def event_router
...
 
+        now = Engine.now
+        record = {'k' => 'v'}
+        assert_rr do
+          mock(filter).filter('test', now, record) { events }
+          event_router.emit('test', now, record)
+        end
+      end
+
+      test 'filter should be called when tag matched and without optimized' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: It's better to check that warning is shown correctly.

**Comment by nurse** [comment]
Just a thought, [stackprof](https://github.com/tmm1/stackprof) is better for performance tuning.

**Comment by ganmacs** [comment]
@tagomoris 

Added these tests.

> only one of some filters has #filter_stream method
filter plugin implements #filter
filter plugin implements #filter_with_time
filters, some implement #filter and others implement #filter_with_time

File: test/test_event_router.rb [hunk 4]
```diff
@@ -247,6 +233,89 @@ def filter_stream(_tag, es)
       end
     end
 
+    sub_test_case 'optimized filter' do
+      setup do
+        @record = { 'k' => 'v' }
+        @now = Engine.now
+      end
+
+      test 'call optimized filter when filter plugin that has #method' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: `#filter` instead of `#method`, right?

File: test/test_event_router.rb [hunk 5]
```diff
@@ -247,6 +233,89 @@ def filter_stream(_tag, es)
...
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter).filter('test', @now, @record) { @record }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test 'call optimized filter when filter plugin that has #filter_method' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: `#filter_with_time` instead of `#filter_method`, right?

File: test/test_event_router.rb [hunk 6]
```diff
@@ -247,6 +233,89 @@ def filter_stream(_tag, es)
...
+
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter).filter_with_time('test', @now, @record) { [time, @record] }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test "don't call optimized filter whenf filter plugins that has #filter_stream" do
```
Comment (tagomoris) [inline test/test_event_router.rb]: typo > `whenf`
Do you mean "when the filter plugin has #filter_stream" ?

**Comment by tagomoris** [comment]
@ganmacs please check and fix the subjects of tests newly added (especially for English syntax etc).

File: test/test_event_router.rb (hunk 4 above)
Comment (ganmacs) [inline test/test_event_router.rb]: fixed.
https://github.com/fluent/fluentd/pull/1145/commits/60dad988dea24b37926dbfa2d739a07f00d69b0f#diff-50a54111d90a980c1d43f9dbfb7632b2R242

File: test/test_event_router.rb (hunk 5 above)
Comment (ganmacs) [inline test/test_event_router.rb]: fixed.
https://github.com/fluent/fluentd/pull/1145/commits/60dad988dea24b37926dbfa2d739a07f00d69b0f#diff-50a54111d90a980c1d43f9dbfb7632b2R251

File: test/test_event_router.rb (hunk 6 above)
Comment (ganmacs) [inline test/test_event_router.rb]: fixed.
https://github.com/fluent/fluentd/pull/1145/commits/60dad988dea24b37926dbfa2d739a07f00d69b0f#diff-50a54111d90a980c1d43f9dbfb7632b2R265

File: test/test_event_router.rb [hunk 7]
```diff
@@ -229,6 +233,89 @@ def event_router
...
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter_with_time).filter_with_time('test', @now, @record) { [@now + 1, @record] }
+          mock(filter).filter('test', @now + 1, @record) { @record }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test "don't call optimized filter whenf only one of some filters has #filter_stream method" do
```
Comment (tagomoris) [inline test/test_event_router.rb]: "whenf"

File: test/test_event_router.rb [hunk 8]
```diff
@@ -229,6 +233,89 @@ def event_router
...
+
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter).filter_stream('test', is_a(OneEventStream)) { OneEventStream.new(@now, @record) }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test 'call optimized filter when filter plugins that have #filter_method and #filter' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: `#filter_method` ?

Comment (tagomoris) [inline test/test_event_router.rb]: "when filter plugins have `#filter_with_time` instead of `#filter`"?

File: test/test_event_router.rb (hunk 7 above)
Comment (tagomoris) [inline test/test_event_router.rb]: "even if just a filter of some filters implements `#filter_stream` method"?

File: test/test_event_router.rb [hunk 9]
```diff
@@ -229,6 +233,89 @@ def event_router
...
+
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter).filter_with_time('test', @now, @record) { [time, @record] }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test "don't call optimized filter when filter plugins that has #filter_stream" do
```
Comment (tagomoris) [inline test/test_event_router.rb]: "when filter plugins have `#filter_stream`"

File: test/test_event_router.rb [hunk 10]
```diff
@@ -229,6 +233,89 @@ def event_router
...
+      test 'call optimized filter when filter plugin that has #filter' do
+        event_router.add_rule('test', filter)
+
+        assert_rr do
+          mock(filter).filter('test', @now, @record) { @record }
+          event_router.emit('test', @now, @record)
+        end
+      end
+
+      test 'call optimized filter when filter plugin that has #filter_with_time' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: "when the filter plugin implements `#filter_with_time` without `#filter_stream`"

File: test/test_event_router.rb [hunk 11]
```diff
@@ -229,6 +233,89 @@ def event_router
       end
     end
 
+    sub_test_case 'optimized filter' do
+      setup do
+        @record = { 'k' => 'v' }
+        @now = Engine.now
+      end
+
+      test 'call optimized filter when filter plugin that has #filter' do
```
Comment (tagomoris) [inline test/test_event_router.rb]: "when the filter plugin implements `#filter` without `#filter_stream`"

File: lib/fluent/plugin/in_dummy.rb [hunk 12]
```diff
@@ -30,6 +30,8 @@ class DummyInput < Input
 
     desc "The value is the tag assigned to the generated events."
     config_param :tag, :string
+    desc "It configures how many events packed in event."
```
Comment (tagomoris) [inline lib/fluent/plugin/in_dummy.rb]: "The number of events in event stream of each emits"

Comment (ganmacs) [inline lib/fluent/plugin/in_dummy.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1145/files#diff-cb31b74554c92f794930629e1a0f6bebR33

File: test/test_event_router.rb (hunk 11 above)
Comment (ganmacs) [inline test/test_event_router.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1145/files#diff-50a54111d90a980c1d43f9dbfb7632b2R242

File: test/test_event_router.rb (hunk 10 above)
Comment (ganmacs) [inline test/test_event_router.rb]: Fixed.

https://github.com/fluent/fluentd/pull/1145/files#diff-50a54111d90a980c1d43f9dbfb7632b2R251

File: test/test_event_router.rb (hunk 9 above)
Comment (ganmacs) [inline test/test_event_router.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1145/files#diff-50a54111d90a980c1d43f9dbfb7632b2R265

File: test/test_event_router.rb (hunk 8 above)
Comment (ganmacs) [inline test/test_event_router.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1145/files#diff-50a54111d90a980c1d43f9dbfb7632b2R279

File: test/test_event_router.rb (hunk 7 above)
Comment (ganmacs) [inline test/test_event_router.rb]: Fixed.
https://github.com/fluent/fluentd/pull/1145/files#diff-50a54111d90a980c1d43f9dbfb7632b2R295

**Comment by tagomoris** [comment]
LGTM. Thank you!

**Pull request merged**